engine/src/main/java/io/zeebe/engine/processor/workflow/CatchEventBehavior.java

```diff
@@ -286,4 +274,48 @@ private boolean sendOpenMessageSubscription(
 
     return extractedCorrelationKeys;
   }
+
+  protected static final class CorrelationKeyExtractionResultHandler
```

Comment:
Instead of externalizing the behavior here, we can also create a specific method in `ExpressionProcessor`. Then all expression related logic (here: extract string or long as string) can be in one place and it is easy to have consistent behavior. What do you think?

Reply:
Here I have the strongest objections. 

To be precise, my objections are not about the conversion logic, but about the side effects. 

In the old code, whenever I wanted to get a String, I had to pass in a `BpmnStepContext` and as a side effect an Incident might be raised. I was very surprised by that. 

If I added a new method to expression processor, then this method would need to get a `MessageCorrelationKeyContect` and again the side effect would be a surprise. Plus, I doubt this method can be reused in many places. But I might be wrong.

I extracted it into a class, because it made it easier for me to write the test. Before that I had it as a methd in `CatchEventBehavior`. But when I wrote the unit test, I had to make so many things just to call the constructor, that the dedicated class clearly was the better options.

I don't care much about where the class lives. I hid it here, because I didn't think it could be resued. If it can be reused, I would probably move it out and alongside `ExpressionResultTypeVerifyingIncidentRaisingHandler`. 

To come back to your initital intent: I think the `ExpressionProcessor` would make more sense, if we can factor out the side effects (raising incidents, or exceptions). So in an ideal world it would return an `Either<Result, ErroMessage>`. But I didn't want to make this change given the pending refactoring.